webui/react/src/hooks/useAsync.ts

```diff
@@ -0,0 +1,55 @@
+import { useCallback, useEffect, useInsertionEffect, useRef, useState } from 'react';
+
+import { Loadable, Loaded, NotLoaded } from 'utils/loadable';
+
+type LoadablePromiser<T> = (canceler: AbortController) => Promise<T | Loadable<T>>;
+
+/**
+ * A hook that manages the result of an async function. The async function is
+ * called as a render effect with an AbortController which is aborted on
+ * unmount.  While pending, the hook returns `NotLoaded`, and on complete, the
+ * hook returns the return value wrapped in `Loaded`. If the function returns a
+ * `Loadable`, the value isn't wrapped. When any value in the deps array
+ * changes, the function is run again as a render effect. If the deps array
+ * changes before the async function returns, the older call will not trigger an
+ * update on return.
```

Comment:
> If the deps array changes before the async function returns, the older call will not trigger an update on return.
 Do you mean here that we cancel the in-flight promise and then run the function again?

Reply:
No -- You can't really cancel a promise from outside the promise, so any side effects in the async function will still occur. `useAsync` will skip the state to the return value of the function.

Reply:
Ah got it. Maybe a better wording here is that "useAsync will discard the result of the older call".

Reply:
good point, updated!